File: objc/kwestkingdom/src/Bow.m
/**
 * Copyright 2009 David Couzelis
 * 
 * This file is part of "Kwest Kingdom".
 * 
 * "Kwest Kingdom" is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * "Kwest Kingdom" is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with "Kwest Kingdom".  If not, see <http://www.gnu.org/licenses/>.
 */
#import "Bow.h"


typedef enum {
  BOW_HOLD_STATE,
  BOW_AWAY_STATE
} BOW_STATE;


@implementation Bow


- init {
  
  self = [super init];
  
  if (self) {
    
    arrow = nil;
    
    holdAnimation = [[Animation alloc] init];
    [holdAnimation addFrame: getImage(IMG_BOW_HOLD_1)];
    [holdAnimation addFrame: getImage(IMG_BOW_HOLD_2)];
    [holdAnimation addFrame: getImage(IMG_BOW_HOLD_3)];
    [holdAnimation addFrame: getImage(IMG_BOW_HOLD_4)];
    [holdAnimation setLoop: YES];
    [holdAnimation setSpeed: 6];
	
    attackRightAnimation = [[Animation alloc] init];
    [attackRightAnimation addFrame: getImage(IMG_BOW_DRAW_1)];
    [attackRightAnimation addFrame: getImage(IMG_BOW_DRAW_2)];
    [attackRightAnimation addFrame: getImage(IMG_BOW_DRAW_3)];
    [attackRightAnimation setLoop: NO];
    [attackRightAnimation setSpeed: 12];
	
    attackLeftAnimation = [[attackRightAnimation copy] setHorizontalFlip: YES];
    attackDownAnimation = [[attackRightAnimation copy] setRotate: YES];
    attackUpAnimation = [[[attackRightAnimation copy] setHorizontalFlip: YES] setRotate: YES];
    
    [self toAwayState];
    
  }
  
  return self;
  
}


- (void) dealloc {
  [holdAnimation release];
  [attackUpAnimation release];
  [attackDownAnimation release];
  [attackLeftAnimation release];
  [attackRightAnimation release];
  [arrow release];
  [super dealloc];
}


- update {
  [super update];
  [arrow update];
  return self;
}


- draw: (BITMAP *) buffer {
  [super draw: buffer];
  [arrow draw: buffer];
  return self;
}


- setArrow: (Arrow *) anArrow {
  arrow = anArrow;
  return self;
}


- (Arrow *) getArrow {
  return arrow;
}


- setArrowWithX: (int) newX
    andY: (int) newY
    andDirection: (int) aDirection
    andTeam: (int) aTeam
    andWorld: (id<Inhabitable, Targetable, Traversable>) aWorld {
  
  arrow = [[Arrow alloc] init];
  [arrow setTeam: aTeam];
  [arrow setX: newX];
  [arrow setY: newY];
  [arrow setWorld: aWorld];
  [arrow setDirection: aDirection];
  [arrow toHoldState];
  
  return self;
  
}


- toHoldState {
  state = BOW_HOLD_STATE;
  animation = holdAnimation;
  return self;
}


- toAwayState {
  state = BOW_AWAY_STATE;
  animation = nil;
  return self;
}


- toAttackUpState {
  animation = attackUpAnimation;
  [animation reset];
  return self;
}


- toAttackDownState {
  animation = attackDownAnimation;
  [animation reset];
  return self;
}


- toAttackLeftState {
  animation = attackLeftAnimation;
  [animation reset];
  return self;
}


- toAttackRightState {
  animation = attackRightAnimation;
  [animation reset];
  return self;
}


- (BOOL) held {
  if (state != BOW_AWAY_STATE) {
    return YES;
  }
  return NO;
}


@end

File: allegro5/src/Bow.cpp
#include "Arrow.h"
#include "Bow.h"
#include "resources.h"


typedef enum {
  BOW_HOLD_STATE,
  BOW_AWAY_STATE
} BOW_STATE;


Bow::Bow() {
  
  arrow = NULL;
  
  init_anim(&hold_anim, ON, 6);
  add_frame(&hold_anim, IMG("bow_hold_1.bmp"));
  add_frame(&hold_anim, IMG("bow_hold_2.bmp"));
  add_frame(&hold_anim, IMG("bow_hold_3.bmp"));
  add_frame(&hold_anim, IMG("bow_hold_4.bmp"));
  
  init_anim(&attack_right_anim, OFF, 12);
  add_frame(&attack_right_anim, IMG("bow_draw_1.bmp"));
  add_frame(&attack_right_anim, IMG("bow_draw_2.bmp"));
  add_frame(&attack_right_anim, IMG("bow_draw_3.bmp"));
  
  copy_anim(&attack_left_anim, &attack_right_anim);
  attack_left_anim.h_flip = ON;

  copy_anim(&attack_down_anim, &attack_right_anim);
  attack_down_anim.rotate = ON;

  copy_anim(&attack_up_anim, &attack_right_anim);
  attack_up_anim.h_flip = ON;
  attack_up_anim.rotate = ON;

  this->toAwayState();
}


Bow::~Bow() {
  delete arrow;
}


void Bow::update() {
  Sprite::update();

  if (arrow) {
    arrow->update();
  }
}


void Bow::draw(IMAGE * canvas) {
  Sprite::draw(canvas);

  if (arrow) {
    arrow->draw(canvas);
  }
}


void Bow::setArrow(Arrow * anArrow) {
  arrow = anArrow;
}


Arrow * Bow::getArrow() {
  return arrow;
}


void Bow::setArrow(int newX, int newY, int aDirection, int aTeam, World *aWorld) {
  
  arrow = new Arrow();
  arrow->setTeam(aTeam);
  arrow->setX(newX);
  arrow->setY(newY);
  arrow->setWorld(aWorld);
  arrow->setDirection(aDirection);
  arrow->toHoldState();
}


void Bow::toHoldState() {
  state = BOW_HOLD_STATE;
  anim = &hold_anim;
}


void Bow::toAwayState() {
  state = BOW_AWAY_STATE;
  anim = NULL;
}


void Bow::toAttackUpState() {
  anim = &attack_up_anim;
  reset_anim(anim);
}


void Bow::toAttackDownState() {
  anim = &attack_down_anim;
  reset_anim(anim);
}


void Bow::toAttackLeftState() {
  anim = &attack_left_anim;
  reset_anim(anim);
}


void Bow::toAttackRightState() {
  anim = &attack_right_anim;
  reset_anim(anim);
}


bool Bow::held() {
  if (state != BOW_AWAY_STATE) {
    return true;
  }
  return false;
}

